### PR TITLE
Allow custom period for invoice generation.

### PR DESCRIPTION
**Pull request by hgaiser**

### CODE REVIEW

File: tools/src/bin/uurlog/invoice.rs [hunk 1]
```diff
@@ -18,7 +18,19 @@ pub struct InvoiceOptions {
 	/// The period to create an invoice for.
 	#[structopt(long)]
 	#[structopt(value_name = "YYYY[-MM[-DD]]")]
-	period: PartialDate,
+	period: Option<PartialDate>,
+
+	/// Only consider records from this date or later.
```
Comment (de-vri-es) [inline tools/src/bin/uurlog/invoice.rs]:
```suggestion
	/// Only consider hour entries from this date or later.
```

File: tools/src/bin/uurlog/invoice.rs [hunk 2]
```diff
@@ -18,7 +18,19 @@ pub struct InvoiceOptions {
...
+	#[structopt(long)]
+	#[structopt(value_name = "YEAR[-MONTH[-DAY]]")]
+	#[structopt(conflicts_with = "period")]
+	start_date: Option<PartialDate>,
+
+	/// Only consider records from this date or earlier.
```
Comment (de-vri-es) [inline tools/src/bin/uurlog/invoice.rs]:
```suggestion
	/// Only consider hour entries from this date or earlier.
```

File: tools/src/bin/uurlog/invoice.rs [hunk 3]
```diff
@@ -18,7 +18,19 @@ pub struct InvoiceOptions {
 	/// The period to create an invoice for.
 	#[structopt(long)]
 	#[structopt(value_name = "YYYY[-MM[-DD]]")]
-	period: PartialDate,
+	period: Option<PartialDate>,
```
Comment (de-vri-es) [inline tools/src/bin/uurlog/invoice.rs:22]: Hmm, I do kinda want period or start/end date to be mandatory. Making an invoice for the full log is not something that really makes sense.

File: tools/src/bin/uurlog/main.rs [hunk 4]
```diff
@@ -44,6 +44,18 @@ struct ShowOptions {
 	#[structopt(long)]
 	#[structopt(value_name = "YYYY[-MM[-DD]]")]
 	period: Option<PartialDate>,
+
+	/// Only consider records from this date or later.
```
Comment (de-vri-es) [inline tools/src/bin/uurlog/main.rs]:
```suggestion
	/// Only consider hour entries from this date or later.
```

File: tools/src/bin/uurlog/main.rs [hunk 5]
```diff
@@ -44,6 +44,18 @@ struct ShowOptions {
...
+	#[structopt(long)]
+	#[structopt(value_name = "YEAR[-MONTH[-DAY]]")]
+	#[structopt(conflicts_with = "period")]
+	start_date: Option<PartialDate>,
+
+	/// Only consider records from this date or earlier.
```
Comment (de-vri-es) [inline tools/src/bin/uurlog/main.rs]:
```suggestion
	/// Only consider hour entries from this date or earlier.
```

File: tools/src/bin/uurlog/invoice.rs (hunk 3 above)
Comment (de-vri-es) [reply tools/src/bin/uurlog/invoice.rs:22]: How about making start-date require end-date, end-date require start-date and adding `period` and `start-date` to a group to make one of them mandatory?

**Comment by hgaiser** [comment]
How about `required_unless` for start and end date and
`required_unless_one` for period?

**Comment by de-vri-es** [comment]
Sounds good :+1:

**Comment by hgaiser** [comment]
I ended up using `required_unless_one` for all three options, since it made sense to me to have only a start date, but not an end date.

**Comment by de-vri-es** [comment]
> I ended up using `required_unless_one` for all three options, since it made sense to me to have only a start date, but not an end date.

Fair enough, I can accept that :p

**Comment by de-vri-es** [comment]
Except that now you can also only specify an `end_date` but no start date. That seems a bit iffy to me.

**Comment by hgaiser** [comment]
> Except that now you can also only specify an `end_date` but no start date. That seems a bit iffy to me.

I didn't see the harm in that, do you?

**Comment by de-vri-es** [comment]
Yes, you're probably doing something you didn't intend to do.

It might even be nice to take `--until-now` instead of no `--end-date`.

**Pull request merged**